Refactor tests in snapshottable and fix #73625

AndiLi99 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #73625 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

AndiLi99 commented:
/assign @davidz627

AndiLi99 commented:
cla recheck

AndiLi99 commented:
recheck cla

AndiLi99 commented:
recheck cla

AndiLi99 commented:
/check-cla

yuxiangqian commented:
/ok-to-test

msau42 commented:
cc @saikat-royc @mattcary

AndiLi99 commented:
/check-cla

AndiLi99 commented:
/check-cla

xing-yang commented:
/assign

AndiLi99 commented:
I'm encountering a strange 404 error.
```test/e2e/storage/testsuites/snapshottable.go:222
Jun 29 17:23:29.816: Unexpected error:
    <*errors.StatusError | 0xc001ca74a0>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "the server could not find the requested resource",
            Reason: "NotFound",
            Details: {
                Name: "",
                Group: "",
                Kind: "",
                UID: "",
                Causes: [
                    {
                        Type: "UnexpectedServerResponse",
                        Message: "404 page not found",
                        Field: "",
                    },
                ],
                RetryAfterSeconds: 0,
            },
            Code: 404,
        },
    }
    the server could not find the requested resource
occurred
test/e2e/storage/testsuites/snapshottable.go:198```
Here is the line: https://github.com/kubernetes/kubernetes/blob/417bda57cb65e4d89dfa5c11553ba47619a59726/test/e2e/storage/testsuites/snapshottable.go#L197
What could be causing this? Perhaps the CRDs are not being registered before the `ginkgo.BeforeEach`?

xing-yang commented:
Yeah, probably the snapshot CRDs are not registered yet.

Also can you fix the golint error from pull-kubernetes-verify?

AndiLi99 commented:
I just fixed the golint error. Also I added the "[Feature:VolumeSnapshotDataSource]" string to all my it blocks. The lack of it may have been causing the test case to run in the wrong job.

AndiLi99 commented:
this test (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92555/pull-kubernetes-e2e-gce-storage-snapshot/1277695556403073024/) failed because the way the logging framework deals with long test names places a limit on the  nesting of descriptions using ginkgo contexts.

The filename limit is 255 bytes for most filesystems (https://serverfault.com/questions/9546/filename-length-limits-on-linux) and the library concatenates the contexts and its together.

I created an issue to track this #92622

I will reduce the length of the descriptions for this PR.

AndiLi99 commented:
/test pull-kubernetes-e2e-kind 

Is this a flaky test?

AndiLi99 commented:
/test pull-kubernetes-e2e-kind 

`pull-kubernetes-e2e-kind — Pod pending timeout.` what is this error?

AndiLi99 commented:
My tests seem to be timing out, should I label them [Slow] or do we need to increase the timeout or the number of parallel nodes running tests?

Or is there another issue that needs to be resolved?

AndiLi99 commented:
It looks like there is a history of these jobs being very close to the edge of the timeout limit: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-storage-snapshot

Previous job runs have seen it fail or succeed based on timeout with no underlying code changes.

AndiLi99 commented:
/test pull-kubernetes-bazel-test

AndiLi99 commented:
I think I have a better understanding of why the tests keeping timing out and failing.

Prow only runs the tests using the internal drivers found at https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go which involves installing and uninstalling the driver at each test run. Since I am testing on my local cluster using the external method there may be differences in the environment causing the timeout. 

https://github.com/kubernetes-csi/external-snapshotter/pull/329 is related.

AndiLi99 commented:
So this job was already failing sometimes due to timeout https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/92387/pull-kubernetes-e2e-gce-storage-snapshot/1280190077598699521. This job was running code unrelated to this PR. 

We've increased the timeout and hopefully it works now.

/test pull-kubernetes-typecheck

AndiLi99 commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-e2e-kind

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest

AndiLi99 commented:
#92886 flake

AndiLi99 commented:
/retest #92741 flaky

AndiLi99 commented:
ping @msau42

AndiLi99 commented:
hit the wrong button

AndiLi99 commented:
/retest

msau42 commented:
Asking @saikat-royc and @mattcary to review first

AndiLi99 commented:
Thank you @msau42.

Thank you @saikat-royc and @mattcary, the PR is ready for review, all jobs passed earlier. I rebased just now to squash some commits so it is still pending and might flake.

mattcary commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,283 +109,245 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				framework.ExpectError(err) // should fail deletion check
+			})
+
+			ginkgo.It("should create objects with default", func() {
```
what do you mean by "with default" here? Isn't this really just checking that the SnapshotContent and such was created correctly/

AndiLi99 replied:
This is actually the wording from the commit before the previous commit. I interpret it to mean as creating objects with default field values. The intention is to just check the creation is done correctly. I think a better wording may be "should create snapshot objects"

mattcary commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,283 +109,245 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				framework.ExpectEqual(snapshotContentSpec["volumeSnapshotClassName"], vsc.GetName())
+				framework.ExpectEqual(volumeSnapshotRef["name"], vs.GetName())
+				framework.ExpectEqual(volumeSnapshotRef["namespace"], vs.GetNamespace())
+			})
```
Your test ends up creating and extra snapshot compared to the original test, and it also doesn't verify delete/retain works with whatever is happening in the snapshot check---which, granted, shouldn't matter now, but might if something is added.

I wonder if it would be better to pull out the checks in "should create objects with default" into a common function that is run by the delete & retain steps, or put it into the beforeEach step.

AndiLi99 replied:
Please check the next commit. It's confusing because there was more refactoring done after I had added the new test.

AndiLi99 replied:
Now there is a test pattern which handles "retain" vs "delete" deletion policy. For each It statement it should be testing both policies now. 

Regarding the creating an extra snapshot compared to the original test, I don't believe that is the case, see:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L204

mattcary replied:
From what I understand, in the original test (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go), a snapshot is created in TestSnapshottable, which is run once each on the It()s on lines 189 and 204 when testing delete and retain policies (two snapshots total).

If I'm understanding your refactored code, the delete policy is hoisted up to a testpattern. A snapshot is created in CreateSnapshotResource. That happens before each of the It()s on lines 205 and 218 in https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go. So to test both delete and retain, this code is run through two testpatterns (?) and a total of 4 snapshots are created.

Maybe I've missed something?

AndiLi99 replied:
Take a look at https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L321 and https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L264.

I believe both TestSnapshotDeleted and TestSnapshottable create their own snapshot, and snapshot class. So that would be a total of 4 before and after the change.

mattcary replied:
ah, right you are. Thanks.

mattcary reviewed (commented):
Here are my comments on the refactoring commit, I haven't looked at the other one yet.

BenTheElder commented:
https://github.com/kubernetes/kubernetes/pull/92555#issuecomment-651875959
This means the Kubernetes based CI failed to run the job.

Note that if a job isn't marked Required at the bottom of the page it isn't required, only informational.

AndiLi99 commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,283 +109,245 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				StorageClassName: &(sc.Name),
+			}, config.Framework.Namespace.Name)
+
+			framework.Logf("In creating storage class object and pvc object for driver - sc: %v, pvc: %v", sc, pvc)
```
Good point, I will remove the extraneous Logf. For future reference, should I prefer logging with By() or Logf?

mattcary replied:
Seems like By() is the idiomatic choice here. It doesn't look like framework.Logf provides any extra functionality.

AndiLi99 commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,283 +109,245 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+			pvc, err = cs.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+			cleanupSteps = append(cleanupSteps, func() {
+				framework.Logf("deleting claim %q/%q", pvc.Namespace, pvc.Name)
```
Good point, will remove.

AndiLi99 commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,283 +109,245 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
-			if err != nil && !apierrors.IsNotFound(err) {
-				framework.Failf("Error deleting claim %q. Error: %v", l.pvc.Name, err)
+		// This is a bit of a hacky way to do cleanup.
+		// Don't use after each as they won't be run if test is killed.
```
This comment gets changed in the next commit. My work ended up being refactor -> code new test -> refactor to fix failing tests on CI, so it's a bit confusing.

Commit:
Refactor snapshottable and provisioning tests

1. Use ginkgo before each to do common setup
2. Use volume resource to create SC, PV, PVC and handle cleanup
3. Add SnapshotResource to handle creating and cleanup of VS, VSC, VSClass
4. Add test pattern for deletion policy: Delete vs Retain
5. Use test pattern to determine test behaviour
6. Add test pattern for preprovisioned snapshot (not implemented)

These changes are made to consolidate common setup steps and stop resource
leaks by waiting for objects to be deleted.

Commit:
Add data semantics test for volume snapshots

The test steps are as follows:
1. Write some data
2. Take a snapshot
3. Write more data
4. Create a new volume from snapshot
5. Validate data is the old data

AndiLi99 commented:
> [#92555 (comment)](https://github.com/kubernetes/kubernetes/pull/92555#issuecomment-651875959)
> This means the Kubernetes based CI failed to run the job.
> 
> Note that if a job isn't marked Required at the bottom of the page it isn't required, only informational.

Thanks for the clarification.

@mattcary I've rebased to separate the refactor from the new test. It's ready for review now, thanks!

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest

mattcary commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,310 +112,334 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+	if sr.Vs != nil {
+		framework.Logf("deleting snapshot %q/%q", sr.Vs.GetNamespace(), sr.Vs.GetName())
+
+		sr.Vs, err = dc.Resource(SnapshotGVR).Namespace(sr.Vs.GetNamespace()).Get(context.TODO(), sr.Vs.GetName(), metav1.GetOptions{})
```
I think this CleanupResource() organization is a big win. But I want to make sure I understand what's going on.

It looks like the semantics have changed here: if a VolumeSnapshot isn't found, the previous test would fail, whereas this one would continue (see https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L335 I think).

Is that because this may not have been called after WaitForSnapshotReady? Or another reason?

AndiLi99 replied:
The big idea with CleanupResource is that it's purpose should be to clean up as much as it can to prevent resource leaks and be resilient to weird conditions, instead of acting as a way to test the deletion works.

The semantics of deletion should be captured by the deletion test: https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go#L205. If the semantics of the test there has changed that would be an issue. If the volume snapshot was not found in that test then https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go#L281 would throw an error.

mattcary replied:
but line 281 won't fire if the resource is not found---don't we want that to be a test error?

+1 on making CleanupResource robust and doing the testing in the test body, that's very nice organization.

AndiLi99 replied:
I suppose you're right. Although CreateSnapshotResource does wait for the snapshot to be ready before it returns, and will error if it doesnt get created. https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go#L332

I suppose implicitly relying on the creation step to guarantee the snapshot exists prior to deletion would be confusing. It may just be easier for the DeleteAndWaitSnapshot method to not ignore `not found` errors.

AndiLi99 replied:
I changed the behavior of DeleteAndWaitSnapshot to not ignore not found errors, and to return the error instead of failing. The test now checks that the deletion succeeds, so it should catch any not found errors.

mattcary commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -236,7 +237,53 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				command = "cat /mnt/test/data"
+				actualData, err := utils.PodExec(f, restoredPod, command)
+				framework.ExpectNoError(err)
+				framework.ExpectEqual(actualData, "hello world")
```
hello world vs junkdata?

At any rate better to put the data written as a string constant.

AndiLi99 replied:
Right the data written into the pvc afterward is "junkdata". The expected value should be "hello world", as that is what is there when the snapshot was taken. I agree it should probably be a string constant.

mattcary replied:
Ah, yes, "hello world" comes from the init function. Yes, definitely a constant should be used to keep things consistent.

AndiLi99 replied:
I've added a commit to add a string constant.

Commit:
Change expected content to string constant. Cleanup DeleteAndWaitSnapshot

Use the correct format specifier

AndiLi99 commented:
/retest

saikat-royc commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,310 +113,386 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+			})
+
+			ginkgo.It("should delete the VolumeSnapshotContent according to its deletion policy", func() {
+				err = DeleteAndWaitSnapshot(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
```
For a retain test case wont this function return err != nil ? line 210 checks unconditionally ExpectNoError?

saikat-royc replied:
to me this looks simpler (because DeleteAndWaitSnapshot already calls WaitForGVRDeletion()): 

err = DeleteAndWaitSnapshot(...)
switch (policy) {
case delete: framework.ExpectNoError(err)
case retain: framework.ExpectError(err)
}

saikat-royc commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,310 +113,386 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+					Name:     vs.GetName(),
+				}
+
+				restoredPVC.Spec.DataSource = dataSourceRef
```
nits: separate dataSourceRef variable not needed.

AndiLi99 replied:
Good point

saikat-royc commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,310 +113,386 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				ginkgo.By("modifying the data in the source PVC")
+
+				command := fmt.Sprintf("echo '%s' > /mnt/test/data", modifiedMntTestData)
+				RunInPodWithVolume(cs, pvc.Namespace, pvc.Name, "pvc-snapshottable-data-tester", command, config.ClientNodeSelection)
```
Cant we use the pod "pvc-snapshottable-tester" created in init() to modify the contents of the PV?
Here we are starting yet another pod "pvc-snapshottable-data-tester" to modify the pv contents.
If we re-use the "pvc-snapshottable-tester" the test timings would be less. Thoughts?

mattcary replied:
Is there an easy way to talk to the process in that init pod? IIUC it needs to wait until after the snapshot is made before modifying the volume. 

Your point about reducing the timing is good, though, considering all the timeouts we've been having recently. Maybe it would be worth it to create some utility pod infrastructure that could easily communicate back and forth with the test to avoid starting up extra pods.

(all of that out of scope for this PR of course :-)

msau42 replied:
We have some tests that do a Pod exec in order to modify data. That may be a method to consider if we don't want to launch a new pod.

AndiLi99 replied:
The pod "pvc-snapshottable-tester" created in init() is destroyed in init(), although there's no reason we can't keep it around so that we can reuse it. It looks like it currently takes ten seconds for a pod to become ready, and we would save ten seconds.

I was having trouble finding a way to execute commands in an existing pod, perhaps I can use this https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L500 and just ignore the error because I'm not looking for a string in output.

saikat-royc commented on test/e2e/storage/testsuites/snapshottable.go:
```diff
@@ -109,310 +113,386 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+		// We instead dynamically take a snapshot and create another snapshot using
+		// the first snapshot's snapshot handle.
+		ginkgo.Skip("Preprovisioned test not implemented")
+		ginkgo.By("taking a snapshot with deletion policy retain")
```
Don't think these are needed. "not implemented" is good enough. We can add details when the test is added.

AndiLi99 replied:
True these aren't needed, but it's a scaffolding to remember the steps.

saikat-royc commented:
Mostly looks good with minor comments from my side. Good stuff with creating a Create/Cleanup Resource for the snapshots and re-using for each It.

mattcary commented:
/lgtm, thanks for bearing with the iterations Andi.

msau42 commented:
/retest
/lgtm
/approve

AndiLi99 commented:
Thanks for the reviews!

msau42 commented:
/priority important-soon

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest

AndiLi99 commented:
/retest
